File: Sources/ClipStackCore/ClipItem.swift
import Foundation
import CryptoKit

/// What kind of pasteboard payload a history entry holds.
public enum ClipKind: String, Codable, Equatable {
    case text   // plain text, optionally with RTF/HTML rich flavors
    case image  // a bitmap, persisted as a PNG file next to the history JSON
    case file   // one or more file URLs copied from Finder etc.
}

/// One clipboard history entry.
///
/// Image bytes are NOT stored inside this struct (they would bloat the JSON
/// history file); instead `imageFileName` points to a PNG inside the store's
/// `images/` directory.
public struct ClipItem: Codable, Identifiable, Equatable {
    public let id: UUID
    public let kind: ClipKind
    /// Stable content fingerprint used for de-duplication across re-copies.
    public let contentHash: String

    public var plainText: String?
    public var rtfData: Data?
    public var htmlData: Data?

    public var imageFileName: String?
    public var imagePixelWidth: Int?
    public var imagePixelHeight: Int?

    /// Absolute POSIX paths for `.file` entries.
    public var fileURLs: [String]?

    public var createdAt: Date
    public var pinned: Bool
    /// Localized name of the frontmost app when the copy happened.
    public var sourceApp: String?
    /// Approximate payload size in bytes (for display only).
    public var byteSize: Int

    public init(
        id: UUID = UUID(),
        kind: ClipKind,
        contentHash: String,
        plainText: String? = nil,
        rtfData: Data? = nil,
        htmlData: Data? = nil,
        imageFileName: String? = nil,
        imagePixelWidth: Int? = nil,
        imagePixelHeight: Int? = nil,
        fileURLs: [String]? = nil,
        createdAt: Date = Date(),
        pinned: Bool = false,
        sourceApp: String? = nil,
        byteSize: Int = 0
    ) {
        self.id = id
        self.kind = kind
        self.contentHash = contentHash
        self.plainText = plainText
        self.rtfData = rtfData
        self.htmlData = htmlData
        self.imageFileName = imageFileName
        self.imagePixelWidth = imagePixelWidth
        self.imagePixelHeight = imagePixelHeight
        self.fileURLs = fileURLs
        self.createdAt = createdAt
        self.pinned = pinned
        self.sourceApp = sourceApp
        self.byteSize = byteSize
    }

    /// Single-line summary shown in list rows and the status menu.
    public var previewLine: String {
        switch kind {
        case .text:
            let flattened = (plainText ?? "")
                .replacingOccurrences(of: "\n", with: " ⏎ ")
                .trimmingCharacters(in: .whitespaces)
            return flattened.isEmpty ? L("empty_text") : String(flattened.prefix(90))
        case .image:
            var dims = ""
            if let w = imagePixelWidth, let h = imagePixelHeight { dims = "\(w)×\(h) " }
            return "\(L("kind_image")) \(dims)(\(ByteFormat.string(byteSize)))"
        case .file:
            let names = (fileURLs ?? []).map { URL(fileURLWithPath: $0).lastPathComponent }
            if names.count == 1 { return names[0] }
            return String(format: L("files_count_prefix"), names.count) + names.prefix(3).joined(separator: ", ")
        }
    }

    /// Text used by the switcher's search filter.
    public var searchText: String {
        switch kind {
        case .text: return plainText ?? ""
        case .file: return (fileURLs ?? []).joined(separator: "\n")
        case .image: return "image 图片 \(L("kind_image")) " + (sourceApp ?? "")
        }
    }

    // MARK: content fingerprints

    public static func hashOf(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
    public static func hashOfText(_ s: String) -> String { "t-" + hashOf(Data(s.utf8)) }
    public static func hashOfImage(_ d: Data) -> String { "i-" + hashOf(d) }
    public static func hashOfFiles(_ paths: [String]) -> String {
        "f-" + hashOf(Data(paths.joined(separator: "\u{0}").utf8))
    }
}

public enum ByteFormat {
    public static func string(_ n: Int) -> String {
        let f = ByteCountFormatter()
        f.countStyle = .file
        return f.string(fromByteCount: Int64(n))
    }
}

File: windows/ClipStack.Win/ClipItem.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Serialization;

namespace ClipStack.Win;

public enum ClipKind { Text, Image, File }

/// <summary>
/// One clipboard history entry. Mirrors the macOS ClipStackCore model:
/// image bytes live as PNG files next to the history JSON, not inline.
/// </summary>
public class ClipItem
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public ClipKind Kind { get; set; }
    public string ContentHash { get; set; } = "";
    public string? PlainText { get; set; }
    public string? ImageFileName { get; set; }
    public int? ImagePixelWidth { get; set; }
    public int? ImagePixelHeight { get; set; }
    public List<string>? FilePaths { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.Now;
    public bool Pinned { get; set; }
    public string? SourceApp { get; set; }
    public long ByteSize { get; set; }

    [JsonIgnore]
    public string PreviewLine
    {
        get
        {
            switch (Kind)
            {
                case ClipKind.Text:
                    var flat = (PlainText ?? "").Replace("\r", "").Replace("\n", " ⏎ ").Trim();
                    if (flat.Length == 0) return Strings.T("empty_text");
                    return flat.Length > 90 ? flat[..90] : flat;
                case ClipKind.Image:
                    var dims = ImagePixelWidth.HasValue ? $"{ImagePixelWidth}×{ImagePixelHeight} " : "";
                    return $"{Strings.T("image_word")} {dims}({ByteSize / 1024} KB)";
                default:
                    var names = (FilePaths ?? new()).Select(Path.GetFileName).ToList();
                    if (names.Count == 1) return names[0] ?? "";
                    return string.Format(Strings.T("files_prefix"), names.Count) + string.Join(", ", names.Take(3));
            }
        }
    }

    [JsonIgnore]
    public string SearchText => Kind switch
    {
        ClipKind.Text => PlainText ?? "",
        ClipKind.File => string.Join("\n", FilePaths ?? new()),
        _ => "image 图片 " + (SourceApp ?? ""),
    };

    public static string HashOf(byte[] data) =>
        Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();

    public static string HashOfText(string s) => "t-" + HashOf(Encoding.UTF8.GetBytes(s));
    public static string HashOfImage(byte[] png) => "i-" + HashOf(png);
    public static string HashOfFiles(IEnumerable<string> paths) =>
        "f-" + HashOf(Encoding.UTF8.GetBytes(string.Join("\0", paths)));
}
